### PR TITLE
Adding virtual-devices yang models

### PR DESCRIPTION
**Pull request by 5abeel**
Introducing a virtual-devices yang model
Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

### CODE REVIEW

**Comment by 5abeel** [comment]
Model tree view

```
module: openconfig-virtual-devices
  +--rw virtual-devices
     +--rw virtual-device* [device-id]
        +--rw device-id    -> ../config/device-id
        +--rw config
        |  +--rw device-id                virtual-device-id
        |  +--rw type                     identityref
        |  +--rw description?             string
        |  +--rw enabled?                 boolean
        |  +--rw instance?                string
        |  +--rw device-version?          string
        |  +--rw model-version?           string
        |  +--rw bandwidth-limit?         uint32
        |  +--rw requests-limit?          uint32
        |  +--rw host-virtual-device?     string
        |  +--rw host-virtual-function?   string
        +--ro state
           +--ro device-id          virtual-device-id
           +--ro type               identityref
           +--ro description?       string
           +--ro enabled?           boolean
           +--ro instance?          string
           +--ro device-version?    string
           +--ro model-version?     string
           +--ro bandwidth-limit?   uint32
           +--ro requests-limit?    uint32
           +--ro admin-status       enumeration
           +--ro oper-status        enumeration
           +--ro last-change?       oc-types:timeticks64
           +--ro counters
              +--ro in-octets?      oc-yang:counter64
              +--ro in-pkts?        oc-yang:counter64
              +--ro in-discards?    oc-yang:counter64
              +--ro out-octets?     oc-yang:counter64
              +--ro out-pkts?       oc-yang:counter64
              +--ro out-discards?   oc-yang:counter64
              +--ro last-clear?     oc-types:timeticks64
```

File: release/models/virtual-devices/openconfig-virtual-devices.yang [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+
+  contact
+    "IPDK Admin
+    ipdkadmin@gmail.com";
+
+  description
+    "Model for managing virtual devices added/deleted
+    into compute instances such as VMs or whole hosts
+
+    Redistribution and use in source and binary forms, with or without
```
Comment (Namrata-intel) [inline release/models/virtual-devices/openconfig-virtual-devices.yang]: should this license info go on top?

**Review by Namrata-intel — APPROVED** [review]
Please check the license. Rest looks good to me.

File: release/models/virtual-devices/openconfig-virtual-devices.yang (hunk 1 above)
Comment (5abeel) [reply release/models/virtual-devices/openconfig-virtual-devices.yang]: Added license info

**Review by dandaly — APPROVED** [review]
This looks good to start, we also need to represent the queues and device capabilities such as RSS, TSO, etc.

File: release/models/virtual-devices/openconfig-virtual-device-types.yang [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+  prefix "oc-vdevice-types";
+
+  import openconfig-extensions { prefix oc-ext; }
+
+  organization "IPDK - Infrastructure Programming Development Kit";
+
+  contact
+    "IPDK Admin
+    ipdkadmin@gmail.com";
```
Comment (mestery) [inline release/models/virtual-devices/openconfig-virtual-device-types.yang:18]: Does this email address actually exist? Who has control of this address?

File: release/models/virtual-devices/openconfig-virtual-devices.yang [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+  }
+
+  grouping vdevice-model-version {
+    description
+      "Reusable definition for a reference to a virtual device model version";
+
+      leaf model-version {
+        type string;
+        description
+          "Version number of config & oper objects";
```
Comment (mestery) [inline release/models/virtual-devices/openconfig-virtual-devices.yang]: Maybe replace `&` with the word `and` and `open` with `operational`?

**Review by mestery — COMMENTED** [review]
Since I see some very basic statistic here, the authors of this PR should look at https://github.com/opiproject/opi-prov-life/issues/92 and work there to try to align some of these common metrics. The basic starting point here about device statistics makes sense, but can be fleshed out more.

File: release/models/virtual-devices/openconfig-virtual-device-types.yang (hunk 2 above)
Comment (5abeel) [reply release/models/virtual-devices/openconfig-virtual-device-types.yang:18]: @dandaly can you confirm?

Comment (dandaly) [reply release/models/virtual-devices/openconfig-virtual-device-types.yang:18]: Yes this email exists and I monitor it.

File: release/models/virtual-devices/openconfig-virtual-devices.yang [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+    leaf requests-limit {
+      type uint32;
+      units pps;
+      default 0;
+      description
+        "Requests limit for virtual-device";
+    }
+  }
+
+  grouping vdevice-host-virtual-ref {
```
Comment (jamescchoi) [inline release/models/virtual-devices/openconfig-virtual-devices.yang:315]: At some point, we will need to accomodate hstid:pfid:vfid:ftype.
We will need to clarify what the format of the strings should be.
OK with committing for now.

File: release/models/virtual-devices/openconfig-virtual-devices.yang [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+      }
+
+      container config {
+        description
+          "Configurable items at the global level";
+        oc-ext:telemetry-on-change;
+
+        uses vdevice-config;
+        uses vdevice-limits;
+        uses vdevice-host-virtual-ref;
```
Comment (jamescchoi) [inline release/models/virtual-devices/openconfig-virtual-devices.yang:367]: Should this be under state as well?

File: release/models/virtual-devices/openconfig-virtual-devices.yang (hunk 3 above)
Comment (5abeel) [reply release/models/virtual-devices/openconfig-virtual-devices.yang]: Ack. Fixed.

File: release/models/virtual-devices/openconfig-virtual-devices.yang [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+      default "true";
+      description
+        "This leaf contains the configured, desired state of the
+        virtual device.";
+    }
+
+    leaf instance {
+      type string;
+      description
+        "Instance to associate with this device";
```
Comment (jamescchoi) [inline release/models/virtual-devices/openconfig-virtual-devices.yang:104]: At some point, we would want to describe more on how to use this field.

File: release/models/virtual-devices/openconfig-virtual-devices.yang [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+            "Down due to state of lower-layer interface(s).";
+        }
+      }
+      mandatory true;
+      description
+        "The current operational state of the virtual device.";
+      oc-ext:telemetry-on-change;
+    }
+
+    leaf last-change {
```
Comment (jamescchoi) [inline release/models/virtual-devices/openconfig-virtual-devices.yang:192]: If not absolutely needed, can we leave this out for now?

File: release/models/virtual-devices/openconfig-virtual-devices.yang [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+
+      leaf out-discards {
+        type oc-yang:counter64;
+        description
+          "The number of outbound packets that were chosen to be
+          discarded even though no errors had been detected to
+          prevent their being deliverable to the infrastructure.";
+      }
+
+      leaf last-clear {
```
Comment (jamescchoi) [inline release/models/virtual-devices/openconfig-virtual-devices.yang:281]: Can we remove this, unless this is required?

File: release/models/virtual-devices/openconfig-virtual-devices.yang (hunk 7 above)
Comment (dandaly) [reply release/models/virtual-devices/openconfig-virtual-devices.yang:192]: This is needed for alignment with interfaces

**Pull request merged**